### ISSUE TITLE
Jittery lines in `textri()`

### ISSUE THREAD

**Issue by borbware** [nesbox/TIC-80#1644]
This description of a persisting textri issue was posted previously by StinkerB06 in the closed PICO-8 SSPR() function issue:

The `textri()` method isn't really good at all:
![textri() texture warp](https://user-images.githubusercontent.com/30059831/67682723-3b0cbf80-f94d-11e9-945c-741b6648fea0.gif)

Look at this! Notice the jittery lines on the lower-right `textri()`, and the occasional "glitchy" pixels.

```lua
W,H=16,16

function TIC()

	if btn(0) then H=H-1 end
	if btn(1) then H=H+1 end
	if btn(2) then W=W-1 end
	if btn(3) then W=W+1 end
	
	cls()
	textri(0,0,W,0,0,H,8,0,24,0,8,16,false,14)
	textri(W,0,0,H,W,H,24,0,8,16,24,16,false,14)
	
end
```
Please correct this code if needed!

_Originally posted by @StinkerB06 in https://github.com/nesbox/TIC-80/issues/952#issuecomment-546955695_

**Comment by joshgoebel** [comment]
_Oh, you're drawing two triangles and expecting them to fit together perfectly._  I'm personally not sure this is a bug rather than a simple deficiency.

Here is another "broken" example drawing lines:

```
	line(0,0,100,40,9)
	line(1,1,101,41,5)
```

The lines do not "fit together" properly at all.  In a perfect world the lines would not overlap, but they do because our algorithm is only trying to draw the best single line it can, not guarantee that two lines draw next to each other won't interact poorly.

**Comment by borbware** [comment]
The problem isn't in the intersection of the triangles, but rather in how the texture gets drawn in some conditions. As the gif below illustrates, _the whole_ lower triangle is drawn with wobbly lines, but the upper one is not. I doubt this is how it's _supposed_ to work!

![video13](https://user-images.githubusercontent.com/19384640/148929976-0164c31f-8a9e-41e4-b222-526c834a58fe.gif)

**Comment by pke1029** [comment]
The jittery lines are a big problem for me whenever I try to draw any non-square sprite. If this is intended for textri, then we need another "texquad" function for drawing sprites onto arbitrary quads without the defect. This was suggested in issue #952 but was told to use the textri function instead. 

This might be a separate issue but since @joshgoebel brought it up here, I thought I would mention that textri and tri disagree even with the same coordinates. The red triangle is drawn using tri and blue is textri. 

![video1](https://user-images.githubusercontent.com/27065573/148939154-86422a0c-e3ff-409a-a447-208566e37a46.gif)

```
x0, y0 = 80, 50
x1, y1 = 50, 80
x2, y2 = 80, 80
k = 100
poke(0x4000, 10) -- textri color

function TIC()

	if btn(0) then y2=y2-1 end
	if btn(1) then y2=y2+1 end
	if btn(2) then x2=x2-1 end
	if btn(3) then x2=x2+1 end

	cls()
	print('tri - textri', 40, 40)
	tri(x0, y0, x1, y1, x2, y2, 2)
	textri(x0, y0, x1, y1, x2, y2, 0, 0, 0, 0, 0, 0)
	
	print('textri - tri', 40+k, 40)
	textri(x0+k, y0, x1+k, y1, x2+k, y2, 0, 0, 0, 0, 0, 0)
	tri(x0+k, y0, x1+k, y1, x2+k, y2, 2)
	
end
```

**Comment by joshgoebel** [comment]
> The problem isn't in the intersection of the triangles, but rather in how the texture gets drawn in some conditions.

Thanks for clarifying - I think I missed that it was just one half. :-)  I agree that's "weird".

**Comment by joshgoebel** [comment]
This is almost certainly our use of floats to translate coordinates... and it's an off-by-one rounding error for some of the calculations:

https://github.com/nesbox/TIC-80/blob/main/src/core/draw.c#L773

No idea how to fix though since most of the time this skewing of perspective seems to be EXPECTED.  

---

I guess the rotations behavior of `texttri` is key otherwise it could have been written much simpler to just copy the spritesheet x,y same as the onscreen x,y, but with an offset?

**Comment by joshgoebel** [comment]
Ah you're just seeing the "jaggies" in the texturing... the top half has a straight edge on the left... so all the offsets work out perfectly... the bottom half has a jagged left edge and therefore some of those pixels start off "off by one" so the entire line is off by one... I feel like I understand it in my head, but still no idea how to fix or if it's easily fixable.

**Comment by eggbanana173** [comment]
no se del tema pero sospecho que el problema puede venir de la forma de que TIC 80 dibuje la textura

**Comment by nesbox** [comment]
As part of the issue, I'm trying to fix the `tri()` function to work with subpixel precision.
![video18](https://user-images.githubusercontent.com/1101448/149818162-267c8dfd-bd78-42a8-9111-74a114194dbc.gif)

Seems, we have to fix all the drawing functions in the same manner.

**Comment by joshgoebel** [comment]
>  to work with subpixel precision.

Could we get a longer explanation of what this means in practice?  Are you talking about rendering to a larger pixel grid and downsampling?

**Comment by nesbox** [comment]
> Are you talking about rendering to a larger pixel grid and downsampling?

No, all the drawing functions now use `int` instead `float` or better `double`, as result we can't do smooth animations and have jittery lines in the `textri()`.

**Comment by joshgoebel** [comment]
Ah, you're just talking about fixing the rounding then.  I'd worry that might be a lot slower on any hardware that doesn't have CPU ops for floating point instructions... not sure that will show up for you if you're benchmarking on a modern Intel or Apple chip. Although perhaps for all the hardware we target this isn't actually a problem?

It sure looks pretty.

**Comment by joshgoebel** [comment]
I guess you'd really only need this for the `line` primitives though (for flood fills)? (textures are a harder problem)

**Comment by nesbox** [comment]
Also found interesting rules for Direct3D 11 
https://docs.microsoft.com/en-us/windows/win32/direct3d11/d3d10-graphics-programming-guide-rasterizer-stage-rules
### Triangle Rasterization Rules (Without Multisampling)
Any pixel center which falls inside a triangle is drawn; a pixel is assumed to be inside if it passes the top-left rule. The top-left rule is that a pixel center is defined to lie inside of a triangle if it lies on the top edge or the left edge of a triangle.

Where:

A top edge, is an edge that is exactly horizontal and is above the other edges.
A left edge, is an edge that is not exactly horizontal and is on the left side of the triangle. A triangle can have one or two left edges.
The top-left rule ensures that adjacent triangles are drawn once.

This illustration shows examples of pixels that are drawn because they either lie inside a triangle or follow the top-left rule.
![image](https://user-images.githubusercontent.com/1101448/149822018-8cd98496-5159-441c-9511-8b364ff59e9d.png)

**Comment by joshgoebel** [comment]
>I guess you'd really only need this for the line primitives though 

I take that back if you wanted to do something like Direct 3d... you can't just know "how to draw a single line" right you have to know which side of a triangle you're on... although that's only to avoid drawing overlapping pixels.  I'm not sure we care about a little overlap as long as there are no "holes"... I'm not sure if simple rounding accomplishes that or not though.

Is the video you showed earlier just using floats?

**Comment by nesbox** [comment]
> Is the video you showed earlier just using floats?

Yes, you can see the difference between old and new tri functions (left - old, right - new)
![video20](https://user-images.githubusercontent.com/1101448/149884526-8674f24e-21d7-4b2e-aaa5-4b29416aeb2a.gif)

**Comment by joshgoebel** [comment]
Is it less jittery?  It's hard to see when its moving :-)

**Comment by nesbox** [comment]
Some progress with the new triangles rasterizer

![video22](https://user-images.githubusercontent.com/1101448/150594279-933ca89e-4ab2-4885-81e2-a5bd721d29a3.gif)

**Comment by nesbox** [comment]
Another comparison

Old rasterizer
![old](https://user-images.githubusercontent.com/1101448/150672145-865e6e81-8ab1-459e-bb85-e3ae9ce2319b.gif)

New rasterizer
![new](https://user-images.githubusercontent.com/1101448/150672141-a100611b-7994-4879-9d2c-a40f30f14e5a.gif)

**Comment by joshgoebel** [comment]
Looks good.  Where are the benchmarks? ;-)

**Comment by nesbox** [comment]
It's slow at the moment but very accurate, will optimize it later :)

**Comment by nesbox** [comment]
New triangle and line rasterizers have been added here e463cc68527df10153107d84a762ef30a8f11c29 , they are not as fast as the previous ones, but much more accurate, and the `tri()` `trib()` `textri()` `line()` API now uses floats as input.

Simple code to check:
``` lua
function TIC()
 w,h=mouse()

 cls()
 textri(0,0,w,0,0,h,8,0,24,0,8,16)
 textri(0,h,w,0,w,h,8,16,24,0,24,16)
end
```
![video26](https://user-images.githubusercontent.com/1101448/151302320-f172ec80-fefc-481b-ba73-a8fcd9dd6c14.gif)

This also fix for #1059

**Comment by joshgoebel** [comment]
> and the tri() trib() textri() line() API now uses floats as input.

@nesbox These is the kind fo binary API breaking changes you can't anymore make once we merge the WASM branch.  You can't just change the C API willy-nilly to take different arguments.

I will update the PR.
